// sfdmu/lib/modules/app/appSfdmuRunApp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const appJsonMessages_1 = require("./appJsonMessages");
const appConsoleUxLogger_1 = require("./appConsoleUxLogger");
const path = require("path");
const runCommandExecutor_1 = require("../commands_processors/runCommandExecutor");
const common_1 = require("../components/common_components/common");
const root = path.resolve(__dirname, '../../../');
const commandMessages = new appJsonMessages_1.default(root, "run");
const resources = new appJsonMessages_1.default(root, "resources");
class AppSfdmuRunApp {
    constructor(args) {
        this.m_flags = {};
        this.m_ux = args.logger || new appConsoleUxLogger_1.default();
        this.argv = args.argv.splice(2);
        this.exportJson = args.exportJson;
        this.commandMessages = args.commandMessages || commandMessages;
        this.resources = args.resources || resources;
        this.exitProcess = args.exitProcess;
        let flags = common_1.Common.parseArgv(...this.argv);
        Object.assign(this.m_flags, flags);
        this.statics = {
            name: "Run",
            plugin: {
                name: "sfdmu",
                root
            }
        };
        this.cmd = {
            statics: this.statics,
            argv: this.argv
        };
    }
    async runCommand() {
        await runCommandExecutor_1.default.execute(this);
    }
}
exports.default = AppSfdmuRunApp;
