// sfdmu/lib/modules/components/common_components/enumerations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DATA_CACHE_TYPES = exports.SPECIAL_MOCK_PATTERN_TYPES = exports.ADDON_EVENTS = exports.MESSAGE_IMPORTANCE = exports.RESULT_STATUSES = exports.API_ENGINE = exports.OPERATION = exports.DATA_MEDIA_TYPE = void 0;
var DATA_MEDIA_TYPE;
(function (DATA_MEDIA_TYPE) {
    DATA_MEDIA_TYPE[DATA_MEDIA_TYPE["Org"] = 0] = "Org";
    DATA_MEDIA_TYPE[DATA_MEDIA_TYPE["File"] = 1] = "File";
})(DATA_MEDIA_TYPE || (exports.DATA_MEDIA_TYPE = DATA_MEDIA_TYPE = {}));
var OPERATION;
(function (OPERATION) {
    OPERATION[OPERATION["Insert"] = 0] = "Insert";
    OPERATION[OPERATION["Update"] = 1] = "Update";
    OPERATION[OPERATION["Upsert"] = 2] = "Upsert";
    OPERATION[OPERATION["Readonly"] = 3] = "Readonly";
    OPERATION[OPERATION["Delete"] = 4] = "Delete";
    OPERATION[OPERATION["DeleteSource"] = 5] = "DeleteSource";
    OPERATION[OPERATION["DeleteHierarchy"] = 6] = "DeleteHierarchy";
    OPERATION[OPERATION["HardDelete"] = 7] = "HardDelete";
    OPERATION[OPERATION["Unknown"] = 8] = "Unknown";
})(OPERATION || (exports.OPERATION = OPERATION = {}));
var API_ENGINE;
(function (API_ENGINE) {
    API_ENGINE[API_ENGINE["DEFAULT_ENGINE"] = 0] = "DEFAULT_ENGINE";
    API_ENGINE[API_ENGINE["REST_API"] = 1] = "REST_API";
    API_ENGINE[API_ENGINE["BULK_API_V1"] = 2] = "BULK_API_V1";
    API_ENGINE[API_ENGINE["BULK_API_V2"] = 3] = "BULK_API_V2";
})(API_ENGINE || (exports.API_ENGINE = API_ENGINE = {}));
var RESULT_STATUSES;
(function (RESULT_STATUSES) {
    RESULT_STATUSES["Undefined"] = "Undefined";
    RESULT_STATUSES["ApiOperationStarted"] = "ApiOperationStarted";
    RESULT_STATUSES["ApiOperationFinished"] = "ApiOperationFinished";
    RESULT_STATUSES["Information"] = "Information";
    RESULT_STATUSES["JobCreated"] = "JobCreated";
    RESULT_STATUSES["BatchCreated"] = "BatchCreated";
    RESULT_STATUSES["DataUploaded"] = "DataUploaded";
    RESULT_STATUSES["InProgress"] = "InProgress";
    RESULT_STATUSES["Completed"] = "Completed";
    RESULT_STATUSES["FailedOrAborted"] = "FailedOrAborted";
    RESULT_STATUSES["ProcessError"] = "ProcessError";
})(RESULT_STATUSES || (exports.RESULT_STATUSES = RESULT_STATUSES = {}));
var MESSAGE_IMPORTANCE;
(function (MESSAGE_IMPORTANCE) {
    MESSAGE_IMPORTANCE[MESSAGE_IMPORTANCE["Silent"] = 0] = "Silent";
    MESSAGE_IMPORTANCE[MESSAGE_IMPORTANCE["Low"] = 1] = "Low";
    MESSAGE_IMPORTANCE[MESSAGE_IMPORTANCE["Normal"] = 2] = "Normal";
    MESSAGE_IMPORTANCE[MESSAGE_IMPORTANCE["High"] = 3] = "High";
    MESSAGE_IMPORTANCE[MESSAGE_IMPORTANCE["Warn"] = 4] = "Warn";
    MESSAGE_IMPORTANCE[MESSAGE_IMPORTANCE["Error"] = 5] = "Error";
})(MESSAGE_IMPORTANCE || (exports.MESSAGE_IMPORTANCE = MESSAGE_IMPORTANCE = {}));
var ADDON_EVENTS;
(function (ADDON_EVENTS) {
    ADDON_EVENTS["none"] = "none";
    ADDON_EVENTS["onBefore"] = "onBefore";
    ADDON_EVENTS["onAfter"] = "onAfter";
    ADDON_EVENTS["onBeforeUpdate"] = "onBeforeUpdate";
    ADDON_EVENTS["onAfterUpdate"] = "onAfterUpdate";
    ADDON_EVENTS["onDataRetrieved"] = "onDataRetrieved";
    ADDON_EVENTS["filterRecordsAddons"] = "filterRecordsAddons";
})(ADDON_EVENTS || (exports.ADDON_EVENTS = ADDON_EVENTS = {}));
var SPECIAL_MOCK_PATTERN_TYPES;
(function (SPECIAL_MOCK_PATTERN_TYPES) {
    SPECIAL_MOCK_PATTERN_TYPES[SPECIAL_MOCK_PATTERN_TYPES["haveAnyValue"] = 0] = "haveAnyValue";
    SPECIAL_MOCK_PATTERN_TYPES[SPECIAL_MOCK_PATTERN_TYPES["missingValue"] = 1] = "missingValue";
})(SPECIAL_MOCK_PATTERN_TYPES || (exports.SPECIAL_MOCK_PATTERN_TYPES = SPECIAL_MOCK_PATTERN_TYPES = {}));
var DATA_CACHE_TYPES;
(function (DATA_CACHE_TYPES) {
    DATA_CACHE_TYPES["InMemory"] = "InMemory";
    DATA_CACHE_TYPES["CleanFileCache"] = "CleanFileCache";
    DATA_CACHE_TYPES["FileCache"] = "FileCache";
})(DATA_CACHE_TYPES || (exports.DATA_CACHE_TYPES = DATA_CACHE_TYPES = {}));
