// sfdmu/lib/addons/components/sfdmu-run/custom/SfdmuRunCustomAddonApiService.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SfdmuRunCustomAddonApiService_instances, _SfdmuRunCustomAddonApiService_pluginJob_get, _SfdmuRunCustomAddonApiService_getPluginTask;
Object.defineProperty(exports, "__esModule", { value: true });
class SfdmuRunCustomAddonApiService {
    constructor(runtime) {
        _SfdmuRunCustomAddonApiService_instances.add(this);
        this.runtime = runtime;
    }
    getPluginRunInfo() {
        return this.runtime.runInfo;
    }
    getProcessedData(context) {
        return __classPrivateFieldGet(this, _SfdmuRunCustomAddonApiService_instances, "m", _SfdmuRunCustomAddonApiService_getPluginTask).call(this, context).processedData;
    }
    log(module, message, messageType, ...tokens) {
        if (typeof message === 'string') {
            this.runtime.logFormatted(module, message, messageType, ...tokens);
        }
        else {
            this.runtime.log(message, messageType == 'JSON' ? messageType : 'OBJECT', ...tokens);
        }
    }
    getPluginJob() {
        return __classPrivateFieldGet(this, _SfdmuRunCustomAddonApiService_instances, "a", _SfdmuRunCustomAddonApiService_pluginJob_get);
    }
    getPluginTask(context) {
        return __classPrivateFieldGet(this, _SfdmuRunCustomAddonApiService_instances, "m", _SfdmuRunCustomAddonApiService_getPluginTask).call(this, context);
    }
}
_SfdmuRunCustomAddonApiService_instances = new WeakSet(), _SfdmuRunCustomAddonApiService_pluginJob_get = function _SfdmuRunCustomAddonApiService_pluginJob_get() {
    return this.runtime.pluginJob;
}, _SfdmuRunCustomAddonApiService_getPluginTask = function _SfdmuRunCustomAddonApiService_getPluginTask(context) {
    return __classPrivateFieldGet(this, _SfdmuRunCustomAddonApiService_instances, "a", _SfdmuRunCustomAddonApiService_pluginJob_get).tasks.find(task => task.sObjectName == context.objectName);
};
exports.default = SfdmuRunCustomAddonApiService;
