// sfdmu/lib/addons/components/sfdmu-run/sfdmuRunAddonTask.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SfdmuRunAddonTask_migrationJobTask, _SfdmuRunAddonTask_sourceTaskData, _SfdmuRunAddonTask_targetTaskData;
Object.defineProperty(exports, "__esModule", { value: true });
const SfdmuRunAddonTaskData_1 = require("./SfdmuRunAddonTaskData");
class SfdmuRunAddonTask {
    constructor(migrationJobTask) {
        _SfdmuRunAddonTask_migrationJobTask.set(this, void 0);
        _SfdmuRunAddonTask_sourceTaskData.set(this, void 0);
        _SfdmuRunAddonTask_targetTaskData.set(this, void 0);
        __classPrivateFieldSet(this, _SfdmuRunAddonTask_migrationJobTask, migrationJobTask, "f");
        __classPrivateFieldSet(this, _SfdmuRunAddonTask_sourceTaskData, new SfdmuRunAddonTaskData_1.default(migrationJobTask.sourceData), "f");
        __classPrivateFieldSet(this, _SfdmuRunAddonTask_targetTaskData, new SfdmuRunAddonTaskData_1.default(migrationJobTask.targetData), "f");
    }
    getTargetCSVFilename(operation, fileNameSuffix) {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").data.getTargetCSVFilename(operation, fileNameSuffix);
    }
    get sourceCsvFilename() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").data.sourceCsvFilename;
    }
    get operation() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").operation;
    }
    get sObjectName() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").sObjectName;
    }
    get targetSObjectName() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").scriptObject.targetObjectName;
    }
    get sourceToTargetRecordMap() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").data.sourceToTargetRecordMap;
    }
    get sourceToTargetFieldNameMap() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").scriptObject.sourceToTargetFieldNameMap;
    }
    get sourceTaskData() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTask_sourceTaskData, "f");
    }
    get targetTaskData() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTask_targetTaskData, "f");
    }
    get processedData() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").processedData;
    }
    get updateMode() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").updateMode == 'forwards' ? 'FIRST_UPDATE' : 'SECOND_UPDATE';
    }
    get fieldsInQuery() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").data.fieldsInQuery;
    }
    get fieldsToUpdate() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").data.fieldsToUpdate;
    }
    get fieldsInQueryMap() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").scriptObject.fieldsInQueryMap;
    }
    get fieldsToUpdateMap() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").scriptObject.fieldsToUpdateMap;
    }
    mapRecords(records) {
        __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").mapRecords(records);
    }
    get tempRecords() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").tempRecords;
    }
    set tempRecords(records) {
        __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").tempRecords = __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").tempRecords || [];
        __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").tempRecords.splice(0, __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").tempRecords.length);
        __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").tempRecords = __classPrivateFieldGet(this, _SfdmuRunAddonTask_migrationJobTask, "f").tempRecords.concat(records);
    }
}
_SfdmuRunAddonTask_migrationJobTask = new WeakMap(), _SfdmuRunAddonTask_sourceTaskData = new WeakMap(), _SfdmuRunAddonTask_targetTaskData = new WeakMap();
exports.default = SfdmuRunAddonTask;
