// sfdmu/lib/addons/components/sfdmu-run/SfdmuRunAddonTaskData.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SfdmuRunAddonTaskData_taskOrgData;
Object.defineProperty(exports, "__esModule", { value: true });
class SfdmuRunAddonTaskData {
    constructor(taskOrgData) {
        _SfdmuRunAddonTaskData_taskOrgData.set(this, void 0);
        __classPrivateFieldSet(this, _SfdmuRunAddonTaskData_taskOrgData, taskOrgData, "f");
    }
    get task() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTaskData_taskOrgData, "f").task;
    }
    get mediaType() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTaskData_taskOrgData, "f").org.media;
    }
    get sObjectName() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTaskData_taskOrgData, "f").task.sObjectName;
    }
    get isSource() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTaskData_taskOrgData, "f").isSource;
    }
    get idRecordsMap() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTaskData_taskOrgData, "f").idRecordsMap;
    }
    get extIdRecordsMap() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTaskData_taskOrgData, "f").extIdRecordsMap;
    }
    get records() {
        return __classPrivateFieldGet(this, _SfdmuRunAddonTaskData_taskOrgData, "f").records;
    }
}
_SfdmuRunAddonTaskData_taskOrgData = new WeakMap();
exports.default = SfdmuRunAddonTaskData;
