// sfdmu/lib/addons/modules/sfdmu-run/custom-addons/package/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DATA_CACHE_TYPES = exports.DATA_MEDIA_TYPE = exports.OPERATION = void 0;
var OPERATION;
(function (OPERATION) {
    OPERATION[OPERATION["Insert"] = 0] = "Insert";
    OPERATION[OPERATION["Update"] = 1] = "Update";
    OPERATION[OPERATION["Upsert"] = 2] = "Upsert";
    OPERATION[OPERATION["Readonly"] = 3] = "Readonly";
    OPERATION[OPERATION["Delete"] = 4] = "Delete";
    OPERATION[OPERATION["DeleteSource"] = 5] = "DeleteSource";
    OPERATION[OPERATION["DeleteHierarchy"] = 6] = "DeleteHierarchy";
    OPERATION[OPERATION["HardDelete"] = 7] = "HardDelete";
    OPERATION[OPERATION["Unknown"] = 8] = "Unknown";
})(OPERATION || (exports.OPERATION = OPERATION = {}));
var DATA_MEDIA_TYPE;
(function (DATA_MEDIA_TYPE) {
    DATA_MEDIA_TYPE[DATA_MEDIA_TYPE["Org"] = 0] = "Org";
    DATA_MEDIA_TYPE[DATA_MEDIA_TYPE["File"] = 1] = "File";
})(DATA_MEDIA_TYPE || (exports.DATA_MEDIA_TYPE = DATA_MEDIA_TYPE = {}));
var DATA_CACHE_TYPES;
(function (DATA_CACHE_TYPES) {
    DATA_CACHE_TYPES["InMemory"] = "InMemory";
    DATA_CACHE_TYPES["CleanFileCache"] = "CleanFileCache";
    DATA_CACHE_TYPES["FileCache"] = "FileCache";
})(DATA_CACHE_TYPES || (exports.DATA_CACHE_TYPES = DATA_CACHE_TYPES = {}));
