// sfdmu/lib/modules/models/common_models/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SuccessExit = exports.CommandAbortedByAddOnError = exports.CommandAbortedByUserError = exports.UnresolvableWarning = exports.CommandExecutionError = exports.CommandInitializationError = exports.OrgMetadataError = void 0;
class OrgMetadataError extends Error {
    constructor(m) {
        super(m);
    }
}
exports.OrgMetadataError = OrgMetadataError;
class CommandInitializationError extends Error {
    constructor(m) {
        super(m);
    }
}
exports.CommandInitializationError = CommandInitializationError;
class CommandExecutionError extends Error {
    constructor(m) {
        super(m);
    }
}
exports.CommandExecutionError = CommandExecutionError;
class UnresolvableWarning extends Error {
    constructor(m) {
        super(m);
    }
}
exports.UnresolvableWarning = UnresolvableWarning;
class CommandAbortedByUserError extends Error {
    constructor(m) {
        super(m);
    }
}
exports.CommandAbortedByUserError = CommandAbortedByUserError;
class CommandAbortedByAddOnError extends Error {
    constructor(m) {
        super(m);
    }
}
exports.CommandAbortedByAddOnError = CommandAbortedByAddOnError;
class SuccessExit extends Error {
    constructor(m) {
        super(m);
    }
}
exports.SuccessExit = SuccessExit;
