// sfdmu/lib/modules/models/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MigrationJob = exports.MigrationJobTask = exports.SObjectDescribe = exports.SFieldDescribe = exports.ScriptOrg = exports.ScriptObjectSet = exports.ScriptObject = exports.ScriptMappingItem = exports.ScriptMockField = exports.Script = exports.OrgInfo = void 0;
__exportStar(require("./common_models/errors"), exports);
__exportStar(require("./common_models/helpers_classes"), exports);
var orgInfo_1 = require("./common_models/orgInfo");
Object.defineProperty(exports, "OrgInfo", { enumerable: true, get: function () { return orgInfo_1.default; } });
var script_1 = require("./script_models/script");
Object.defineProperty(exports, "Script", { enumerable: true, get: function () { return script_1.default; } });
var scriptMockField_1 = require("./script_models/scriptMockField");
Object.defineProperty(exports, "ScriptMockField", { enumerable: true, get: function () { return scriptMockField_1.default; } });
var scriptMappingItem_1 = require("./script_models/scriptMappingItem");
Object.defineProperty(exports, "ScriptMappingItem", { enumerable: true, get: function () { return scriptMappingItem_1.default; } });
var scriptObject_1 = require("./script_models/scriptObject");
Object.defineProperty(exports, "ScriptObject", { enumerable: true, get: function () { return scriptObject_1.default; } });
var scriptObjectSet_1 = require("./script_models/scriptObjectSet");
Object.defineProperty(exports, "ScriptObjectSet", { enumerable: true, get: function () { return scriptObjectSet_1.default; } });
var scriptOrg_1 = require("./script_models/scriptOrg");
Object.defineProperty(exports, "ScriptOrg", { enumerable: true, get: function () { return scriptOrg_1.default; } });
var sfieldDescribe_1 = require("./sf_models/sfieldDescribe");
Object.defineProperty(exports, "SFieldDescribe", { enumerable: true, get: function () { return sfieldDescribe_1.default; } });
var sobjectDescribe_1 = require("./sf_models/sobjectDescribe");
Object.defineProperty(exports, "SObjectDescribe", { enumerable: true, get: function () { return sobjectDescribe_1.default; } });
var migrationJobTask_1 = require("./job_models/migrationJobTask");
Object.defineProperty(exports, "MigrationJobTask", { enumerable: true, get: function () { return migrationJobTask_1.default; } });
var migrationJob_1 = require("./job_models/migrationJob");
Object.defineProperty(exports, "MigrationJob", { enumerable: true, get: function () { return migrationJob_1.default; } });
